src/plugins/tonegen/dtmf.c

```diff
@@ -0,0 +1,279 @@
+#include "dtmf.h"
+#include "dbusif.h"
+
+#define TRACE(f, args...) trace_write(trctx, trflags, trkeys, f, ##args)
+#define LOG_CAT "tonegen-dtmf: "
+
+#define MUTE_ON  1
+#define MUTE_OFF 0
+
+struct dtmf {
+    char           symbol;
+    uint32_t       low_freq;
+    uint32_t       high_freq;
+};
+
+
+static struct dtmf dtmf_defs[DTMF_MAX] = {
+    {'0', 941, 1336},
+    {'1', 697, 1209},
+    {'2', 697, 1336},
+    {'3', 697, 1477},
+    {'4', 770, 1209},
+    {'5', 770, 1336},
+    {'6', 770, 1477},
+    {'7', 852, 1209},
+    {'8', 852, 1336},
+    {'9', 852, 1477},
+    {'*', 941, 1209},
+    {'#', 941, 1477},
+    {'A', 697, 1633},
+    {'B', 770, 1633},
+    {'C', 852, 1633},
+    {'D', 941, 1633}
+};
+
+static char   *dtmf_stream = STREAM_DTMF;
+static void   *dtmf_props  = NULL;
+static int     vol_scale   = 100;
+static guint   mute        = MUTE_OFF;
+static guint   tmute_id;
+
+
+static void destroy_callback(void *);
+static void set_mute_timeout(struct ausrv *, guint);
+static gboolean mute_timeout_callback(gpointer);
+static void request_muting(struct ausrv *, dbus_bool_t);
+
+
+
+int dtmf_init(void)
+{
+    return 0;
+}
+
+void dtmf_play(struct ausrv *ausrv, uint type, uint32_t vol, int dur, const char *extra_properties)
+{
+    struct stream *stream = stream_find(ausrv, dtmf_stream);
+    struct dtmf   *dtmf   = dtmf_defs + type;
+    uint32_t       per    = dur;
+    uint32_t       play   = dur > 60000 ? dur - 20000 : dur;
+    int            type_l = TONE_DTMF_L;
+    int            type_h = TONE_DTMF_H;
+    uint32_t       timeout;
+    void          *properties = dtmf_props;
+        
+    if (type >= DTMF_MAX || (dur != 0 && dur < 10000))
+        return;
+
+    if (!dur) {
+        /*
+         * These types will make the DTMF tone as 'indicator'
+         * i.e. the indicator_stop() will work on them
+         */
+        type_l = TONE_DTMF_IND_L;
+        type_h = TONE_DTMF_IND_H;
+
+        per = play = 1000000;
+    }
+
+    if (stream != NULL) {
+        if (!dur) {
+            indicator_stop(ausrv, KILL_STREAM);
+            dtmf_stop(ausrv);
+        }
+    }
+    else {
+        if (extra_properties)
+            properties = stream_merge_properties(dtmf_props, extra_properties);
+
+        stream = stream_create(ausrv, dtmf_stream, NULL, 0,
+                               tone_write_callback,
+                               destroy_callback,
+                               properties,
+                               NULL);
+
+        if (extra_properties)
+            stream_free_properties(properties);
+
+        if (stream == NULL) {
+            N_ERROR(LOG_CAT "%s(): Can't create stream", __FUNCTION__);
+            return;
+        }
+    }
+
+    vol = (vol_scale * vol) / 100;
+
+    tone_create(stream, type_l, dtmf->low_freq , vol/2, per,play, 0,dur);
+    tone_create(stream, type_h, dtmf->high_freq, vol/2, per,play, 0,dur);
+
+    timeout = dur ? dur + (30 * 1000000) : (1 * 60 * 1000000);
```

Comment:
Again here might be useful to have named constants to better understand what this does.